### PR TITLE
Add blog post about v03x/v04x migration

### PR DESCRIPTION
**Pull request by lgierth**
I think this has all the essential info. @RichardLitt @jbenet thoughts?

### CODE REVIEW

**Comment by lgierth** [comment]
I'm not even sure a blog post is the perfect medium, but we can certainly use it to discuss and ship

File: src/9-v04x-migration/index.md [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+IPFS nodes built from the master branch are already part of v04x.
+
+ipfs.io provides two essential services to the IPFS community,
+which are affected by this.
+
+
+## The public gateways at ipfs.io
+
+For the time being, https://ipfs.io uses both v03x and v04x to service requests.
+Which ever responds successfully (2xx or 3xx status code) first,
```
Comment (RichardLitt) [inline src/9-v04x-migration/index.md]: whichever

File: src/9-v04x-migration/index.md [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+gets to serve. The other response is discarded.
```
Comment (RichardLitt) [inline src/9-v04x-migration/index.md]: gets to serve its content

File: src/9-v04x-migration/index.md [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+That also means that if the first response is not successful (4xx/5xx),
+it is discarded and the second response is served, regardless of its status code.
+
+If you wanted to target a specific network, use v03x.ipfs.io or v04x.ipfs.io.
+These domain names will stay around as long as the respective network is
+supported by the public gateway.
+
+Expect the v03x gateway to be **supported until end of April 2016.**
```
Comment (RichardLitt) [inline src/9-v04x-migration/index.md]: until the end

File: src/9-v04x-migration/index.md [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+After that day, ipfs.io will be served by v04x exclusively,
+and v03x.ipfs.io will no longer work.
+
+
+## The default bootstrappers
+
+These are the 8 bootstrap nodes in go-ipfs's default config.
```
Comment (RichardLitt) [inline src/9-v04x-migration/index.md]: Which are? Should they be posted here? How do I see the default nodes?

File: src/9-v04x-migration/index.md [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+Their PeerIDs start with `QmSoL` (for Solarnet).
+All of them use `/tcp/4001` and `/udp/4002/utp`.
+In order to balance them over v03x and v04x,
+a few of them bind v04x to these ports, and a few bind v03x.
+The respective other network is bound to `/tcp/14001` and `/udp/14002/utp`.
+The means that all 8 nodes are still part of both networks,
+but are available at default bootstrappers only to one.
+
+We'll gradually shift the v03x/v04x ratio to v04x.
+Expect the v03x bootstrappers to be **supported until end of April 2016.**
```
Comment (RichardLitt) [inline src/9-v04x-migration/index.md]: until the end

**Comment by RichardLitt** [comment]
Overall, LGTM. However: what does this mean for me? Do I need to update my IPFS daemon? Do I need to update my server that uses IPFS? Something to clarify that would be good, I think.

**Comment by lgierth** [comment]
Another thing that comes to mind is the question of whether v03x receives further patches/updates. It doesn't right now and it's not planned, but we should probably say so. cc @jbenet @whyrusleeping

**Comment by lgierth** [comment]
Addressed the feedback and open questions, and generally improved the thing. It's pretty good now I think.

- Preview: https://ipfs.io/ipfs/QmQUuZSSYJRn8sXBDsUfyoWbkD9BS4W7UNXFsAdtRVTwSy/9-v04x-migration/
- Diff: https://github.com/ipfs/blog/commit/26c22b850e67d9955cdeb317d8fa0406ea84748c

File: src/9-v04x-migration/index.md [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+breadcrumbs:
+  - {name: "9-v04x-migration", link: "./" }
+tags: gateway, bootstrap, infrastructure
+title: Migrating ipfs.io from go-ipfs 0.3.x to 0.4.0
+author: Lars Gierth
+collection: posts
+---
+
+Good news everyone! We are about to release go-ipfs 0.4.0,
+which contains lots of great changes and enhancements.
```
Comment (jbenet) [inline src/9-v04x-migration/index.md]: i think we want a separate post about "the great changes and enhancements". can anyone pick this up? maybe @RichardLitt with help of @whyrusleeping and @diasdavid?

File: src/9-v04x-migration/index.md [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+
+There are breaking changes
+which prevent 0.4.x nodes from communicating with 0.3.x nodes.
+This means that there will be, and already are, two separate networks.
+We'll call these networks v03x and v04x.
+IPFS nodes built from the master branch are already part of v04x.
+The Docker image `ipfs/go-ipfs` is built from master and thus also part of v04x.
```
Comment (jbenet) [inline src/9-v04x-migration/index.md]: we should have the docker image follow the tags i have in `jbenet/go-ipfs`:
- `hub.docker.com/u/ipfs/go-ipfs@release` follows `github.com/ipfs/go-ipfs@release` (0.3.11 now)
- `hub.docker.com/u/ipfs/go-ipfs@latest` follows `github.com/ipfs/go-ipfs@release`  (0.3.11 now)
- `hub.docker.com/u/ipfs/go-ipfs@master` follows `github.com/ipfs/go-ipfs@master` (0.4.0 now)

File: src/9-v04x-migration/index.md [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+
+ipfs.io provides two essential services to the community,
+which are affected by this: the gateway and the default bootstrappers.
```
Comment (jbenet) [inline src/9-v04x-migration/index.md]: "the public http-to-ipfs gateway"

File: src/9-v04x-migration/index.md (hunk 7 above)
Comment (lgierth) [inline src/9-v04x-migration/index.md]: We'll be able to have these only after the next release. It actually had a v0.3.11 tag until I removed it today (ipfs/go-ipfs#1944). The problem is that the dockerfile on 0.3.x just downloads `ipfs_master_linux-386.zip` from gobuilder. So it doesn't actually build an 0.3.x image. We'd have to build the image manually with current master's good dockerfile and push that, which doesn't seem to be allowed for automatic builds.

File: src/9-v04x-migration/index.md [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+
+- [How do I update go-ipfs to 0.4.0?](#how-do-i-update-go-ipfs-to-0-4-0-)
+- [The public gateways at ipfs.io](#the-public-gateways-at-ipfs-io)
+- [The default bootstrappers](#the-default-bootstrappers)
+- [How does this affect me?](#how-does-this-affect-me-)
+
+
+## How do I update go-ipfs to 0.4.0?
+
+You can use the [ipfs-update tool][ipfs-update] to update go-ipfs.
```
Comment (jbenet) [inline src/9-v04x-migration/index.md]: You can: 
- use the ipfs-update tool to update go-ipfs **(recommended)**
- download [the new go-ipfs binaries][go-ipfs-bins] and install them.
- download [the new source](https://github.com/ipfs/go-ipfs) and [build it yourself](https://github.com/ipfs/go-ipfs#build-from-source).

Please note that installing with `go get github.com/ipfs/go-ipfs` does not work at this time. We are experimenting with our [ipfs-based package manager](https://github.com/whyrusleeping/gx).

File: src/9-v04x-migration/index.md [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+
+[ipfs-update]: http://dist.ipfs.io/#ipfs-update
+
+
+## The public gateways at ipfs.io
+
+For the time being, https://ipfs.io uses both v03x and v04x to service requests.
```
Comment (jbenet) [inline src/9-v04x-migration/index.md]: add: "Content available in either network will be served just fine."

File: src/9-v04x-migration/index.md [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+Whichever responds successfully first (2xx or 3xx status code),
+gets to serve its content. The other response is discarded.
+If the first response is not successful (4xx/5xx, connection errors),
+it is discarded and the second response is served, regardless of its status code.
+
+We're using the [multireq proxy][multireq] to accomplish this multiplexing behaviour.
```
Comment (jbenet) [inline src/9-v04x-migration/index.md]: Add:
"`multireq` is a new tool, so if you notice any weirdness with gateway requests, [please let us know](https://github.com/ipfs/infrastructure/issues/)."

File: src/9-v04x-migration/index.md [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+which is part of the gateway.
+
+Expect the **v03x gateway** to be **supported until the end of April 2016.**
+After that day, ipfs.io will be served by v04x exclusively,
+and v03x.ipfs.io will no longer work.
+
+[multireq]: https://github.com/whyrusleeping/multireq
+
+
+## The default bootstrappers
```
Comment (jbenet) [inline src/9-v04x-migration/index.md]: maybe: 

## Solarnet: the default bootstrappers

Comment (lgierth) [inline src/9-v04x-migration/index.md]: Oh yep that's a good occasion to subtly highlight what solarnet actually is :)

File: src/9-v04x-migration/index.md [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+
+
+## How does this affect me?
+
+Check which version of go-ipfs you are running: `ipfs version`
+
+If you're running an 0.3.x node, it won't be able to communicate
+with any node which has updated to 0.4.x, let alone exchange data.
+You can access data added by 0.4.0 nodes using the public gateway.
+Likewise, data added by your 0.3.x node is still available on the public gateway.
```
Comment (jbenet) [inline src/9-v04x-migration/index.md]: If you rely on the public gateways, we ask you to please upgrade to 0.4.0 as soon as you can. You can subscribe [to this issue](https://github.com/ipfs/go-ipfs/issues/2334) to be notified when 0.4.0 is released.

File: src/9-v04x-migration/index.md [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+title: Migrating ipfs.io from go-ipfs 0.3.x to 0.4.0
+author: Lars Gierth
+collection: posts
+---
+
+Good news everyone! We are about to release go-ipfs 0.4.0,
+which contains lots of great changes and enhancements.
+
+There are breaking changes
+which prevent 0.4.x nodes from communicating with 0.3.x nodes.
```
Comment (jbenet) [inline src/9-v04x-migration/index.md]: I think the intro needs some acknowledgement that breaking changes suck, that we take them seriously, and that we've a smooth upgrade path.

> We know breaking changes are painful. We avoid them. In this case, the improvement makes IPFS substantially more upgradable. It's one of those things that "should've been different" from the start. Having the freedom to improve the core protocol is why IPFS is still in alpha phase.
>
> Nevertheless, we take service disruption very seriously even in these early days. We know thousands of developers use IPFS, and hundreds of thousands of people rely on IPFS links to our gateway. So we planned a smooth upgrade path.
>
> The breaking change is at the wire protocol level, which means there will be, and already are, two separate networks. ...

Comment (lgierth) [inline src/9-v04x-migration/index.md]: > I think the intro needs some acknowledgement that breaking changes suck, that we take them seriously, and that we've a smooth upgrade path.

That's excellent and the basic tone that I couldn't come up with :)

File: src/9-v04x-migration/index.md (hunk 13 above)
Comment (jbenet) [inline src/9-v04x-migration/index.md]: If you're running an 0.4.0 node, all is well in the world :sunglasses: You don't need to do anything.

Thank you for bearing with this important change. We seek to provide a smooth transition for everyone, and wish we didn't have to bother you with this at all.

File: src/9-v04x-migration/index.md (hunk 7 above)
Comment (lgierth) [inline src/9-v04x-migration/index.md]: Adding a tiny note about this.

**Comment by lgierth** [comment]
Okay addressed feedback

- preview: https://ipfs.io/ipfs/QmYWYq4z5MspeuDCYAyWqko5sG2z4Bx3FqZvMkVgt8RVQq/9-v04x-migration/
- diff: https://github.com/ipfs/blog/commit/4b188000d6a00a992b2575a0d4452bea2ea9d452

**Comment by jbenet** [comment]
LGTM.

**Comment by jbenet** [comment]
thanks @lgierth!

**Pull request merged**